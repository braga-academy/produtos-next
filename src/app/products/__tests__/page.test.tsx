import { render, screen, fireEvent } from '@testing-library/react'
import { Provider } from 'react-redux'
import { configureStore } from '@reduxjs/toolkit'
import ProductsPage from '../page'
import productsReducer from '@/features/products/store/productsSlice'

// Mock do useToast
jest.mock('@/components/ui/use-toast', () => ({
    useToast: () => ({
        toast: jest.fn(),
    }),
}))

// Mock do dispatch
const mockDispatch = jest.fn()
jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useDispatch: () => mockDispatch,
}))

describe('ProductsPage', () => {
    beforeEach(() => {
        mockDispatch.mockClear()
        jest.useFakeTimers()
    })

    afterEach(() => {
        jest.useRealTimers()
    })

    const createMockStore = (state = {}) => configureStore({
        reducer: {
            products: productsReducer,
        },
        preloadedState: {
            products: {
                items: [],
                filteredItems: [],
                loading: false,
                error: null,
                filters: {
                    search: '',
                    category: '',
                    stockFilter: 'all' as const,
                    minPrice: 0,
                    maxPrice: 0,
                    startDate: '',
                },
                pagination: {
                    page: 1,
                    pageSize: 10,
                    total: 0,
                },
                ...state,
            },
        },
    })

    it('deve corresponder ao snapshot do estado vazio', () => {
        const { container } = render(
            <Provider store={createMockStore()}>
                <ProductsPage />
            </Provider>
        )
        expect(container).toMatchSnapshot()
    })

    it('deve exibir o estado de carregamento', () => {
        render(
            <Provider store={createMockStore({ loading: true })}>
                <ProductsPage />
            </Provider>
        )
        // Verifica se existem elementos de skeleton
        const skeletons = document.querySelectorAll('[data-slot="skeleton"]')
        expect(skeletons.length).toBeGreaterThan(0)
    })

    it('deve exibir mensagem de erro', () => {
        const errorMessage = 'Erro ao carregar produtos'
        render(
            <Provider store={createMockStore({ error: errorMessage })}>
                <ProductsPage />
            </Provider>
        )
        // Busca pelo título do erro
        expect(screen.getByRole('heading', { name: errorMessage })).toBeInTheDocument()
    })

    it('deve exibir produtos quando houver dados', () => {
        const mockProducts = [
            {
                id: '1',
                name: 'Produto Teste',
                price: 99.99,
                category: 'Categoria Teste',
                stock: 10,
                description: 'Descrição Teste',
                createdAt: '2024-03-20T10:00:00.000Z',
                imageUrl: null,
            },
        ]

        render(
            <Provider store={createMockStore({
                items: mockProducts,
                filteredItems: mockProducts,
                pagination: {
                    page: 1,
                    pageSize: 10,
                    total: 1,
                }
            })}>
                <ProductsPage />
            </Provider>
        )

        // Verifica se a tabela está presente
        const table = screen.getByRole('table')
        expect(table).toBeInTheDocument()

        // Verifica se o produto está na tabela
        const rows = screen.getAllByRole('row')
        expect(rows).toHaveLength(2) // Header + 1 produto

        // Verifica o conteúdo da linha do produto
        const productRow = rows[1]
        expect(productRow).toHaveTextContent('Produto Teste')
        expect(productRow).toHaveTextContent('R$ 99,99')
        expect(productRow).toHaveTextContent('Categoria Teste')
        expect(productRow).toHaveTextContent('Descrição Teste')
        expect(productRow).toHaveTextContent('10 unidades')
    })

    it('deve abrir o modal ao clicar em Novo Produto', () => {
        render(
            <Provider store={createMockStore()}>
                <ProductsPage />
            </Provider>
        )

        const newProductButtons = screen.getAllByRole('button', { name: 'Novo Produto' })
        fireEvent.click(newProductButtons[0])

        // Verifica se o modal está aberto
        const modal = screen.getByRole('dialog')
        expect(modal).toBeInTheDocument()
        expect(modal).toHaveAttribute('data-state', 'open')
    })

    it('deve filtrar produtos ao digitar na busca', () => {
        // Mock do dispatch para retornar a ação correta
        mockDispatch.mockImplementation((action) => {
            if (action.type === 'products/setFilters') {
                return { type: 'products/setFilters', payload: { search: 'teste' } }
            }
            return action
        })

        render(
            <Provider store={createMockStore()}>
                <ProductsPage />
            </Provider>
        )

        const searchInput = screen.getByPlaceholderText('Buscar produtos...')
        fireEvent.change(searchInput, { target: { value: 'teste' } })

        jest.advanceTimersByTime(300)

        expect(mockDispatch).toHaveBeenCalledWith({
            type: 'products/setFilters',
            payload: { search: 'teste' },
        })
    })

    it('deve exibir a visualização mobile em telas pequenas', () => {
        // Simula uma tela mobile
        window.innerWidth = 375
        window.dispatchEvent(new Event('resize'))

        const mockProducts = [
            {
                id: '1',
                name: 'Produto Teste',
                price: 99.99,
                category: 'Categoria Teste',
                stock: 10,
                description: 'Descrição Teste',
                createdAt: '2024-03-20T10:00:00.000Z',
                imageUrl: null,
            },
        ]

        render(
            <Provider store={createMockStore({
                items: mockProducts,
                filteredItems: mockProducts,
            })}>
                <ProductsPage />
            </Provider>
        )

        // Verifica se a visualização mobile está presente
        const mobileView = document.querySelector('.md\\:hidden')
        expect(mobileView).toBeInTheDocument()

        // Verifica se o produto está na visualização mobile usando getAllByText
        const productNames = screen.getAllByText('Produto Teste')
        expect(productNames.length).toBeGreaterThan(0)

        // Verifica o preço na visualização mobile
        const prices = screen.getAllByText('R$ 99,99')
        expect(prices.length).toBeGreaterThan(0)
    })
})
