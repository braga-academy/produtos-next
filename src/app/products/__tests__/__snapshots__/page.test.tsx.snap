// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsPage deve corresponder ao snapshot do estado vazio 1`] = `
<div>
  <div
    class="container mx-auto py-10 space-y-6"
  >
    <div
      class="flex justify-between items-center"
    >
      <div
        class="space-y-1"
      >
        <h1
          class="text-3xl font-bold tracking-tight"
        >
          Produtos
        </h1>
        <p
          class="text-sm text-muted-foreground"
        >
          Gerencie seus produtos e estoque
        </p>
      </div>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3 cursor-pointer"
        data-slot="button"
        data-testid="create-product-button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-plus mr-2 h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 12h14"
          />
          <path
            d="M12 5v14"
          />
        </svg>
        Novo Produto
      </button>
    </div>
    <div
      class="space-y-4"
    >
      <div
        class="flex gap-4"
      >
        <div
          class="relative flex-1"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-search absolute left-2 top-2.5 h-4 w-4 text-muted-foreground"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m21 21-4.34-4.34"
            />
            <circle
              cx="11"
              cy="11"
              r="8"
            />
          </svg>
          <input
            class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive pl-8"
            data-slot="input"
            data-testid="search-input"
            placeholder="Buscar produtos..."
            value=""
          />
        </div>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 h-9 px-4 py-2 has-[>svg]:px-3 gap-2"
          data-slot="button"
          data-testid="toggle-filters"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-funnel h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 20a1 1 0 0 0 .553.895l2 1A1 1 0 0 0 14 21v-7a2 2 0 0 1 .517-1.341L21.74 4.67A1 1 0 0 0 21 3H3a1 1 0 0 0-.742 1.67l7.225 7.989A2 2 0 0 1 10 14z"
            />
          </svg>
          Filtros
        </button>
      </div>
    </div>
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm border-dashed"
      data-slot="card"
    >
      <div
        class="px-6 pt-6"
        data-slot="card-content"
      >
        <div
          class="flex flex-col items-center justify-center text-center space-y-4 py-12"
        >
          <div
            class="rounded-full bg-muted p-4"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-shopping-cart h-12 w-12 text-muted-foreground"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="8"
                cy="21"
                r="1"
              />
              <circle
                cx="19"
                cy="21"
                r="1"
              />
              <path
                d="M2.05 2.05h2l2.66 12.42a2 2 0 0 0 2 1.58h9.78a2 2 0 0 0 1.95-1.57l1.65-7.43H5.12"
              />
            </svg>
          </div>
          <div
            class="space-y-2"
          >
            <h3
              class="text-lg font-semibold"
            >
              Nenhum produto cadastrado
            </h3>
            <p
              class="text-sm text-muted-foreground"
            >
              Comece adicionando seu primeiro produto
            </p>
          </div>
          <button
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3 gap-2"
            data-slot="button"
            data-testid="empty-create-button"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-plus h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5 12h14"
              />
              <path
                d="M12 5v14"
              />
            </svg>
            Novo Produto
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
